### PR TITLE
net: fota_download: Add API to cancel downloading

### PR DESCRIPTION
**Pull request by junqingzou**
From customer project, cancelling FOTA download is required.
Affect module (no default in switch)
      subsys\net\lib\aws_fota\src\aws_fota

JIRA reference: NCSIDB-353

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: include/net/fota_download.h [hunk 1]
```diff
@@ -105,6 +107,14 @@ int fota_download_init(fota_download_callback_t client_callback);
 int fota_download_start(const char *host, const char *file, int sec_tag,
 			const char *apn, size_t fragment_size);
 
+/**@brief Cancel FOTA image downloading.
+ *
+ * @retval 0 If FOTA is cancelled successfully.
+ *         -EALREADY If download is not started, aborted or completed.
```
Comment (rlubos) [inline include/net/fota_download.h:113]: I believe this needsa separate `@retval` entry (if we list the possible return values).

File: subsys/net/lib/fota_download/src/fota_download.c [hunk 2]
```diff
@@ -367,9 +367,36 @@ int fota_download_init(fota_download_callback_t client_callback)
...
 	return 0;
 }
 
+int fota_download_cancel(void)
+{
+	int err = 0;
+
+	if (dlc.fd == -1) {
+		/* Download not started, aborted or completed */
+		err = -EALREADY;
```
Comment (rlubos) [inline subsys/net/lib/fota_download/src/fota_download.c:380]: Just print the warning and return early with -EAGAIN, to aviod the extra indentation.

File: subsys/net/lib/fota_download/src/fota_download.c [hunk 3]
```diff
@@ -367,9 +367,36 @@ int fota_download_init(fota_download_callback_t client_callback)
...
+		LOG_WRN("%s invalid state", __func__);
+	} else {
+		err = download_client_disconnect(&dlc);
+		if (err) {
+			LOG_ERR("%s failed to disconnect: %d", __func__, err);
+			return err;
+		}
```
Comment (rlubos) [inline subsys/net/lib/fota_download/src/fota_download.c:387]: Please add a blank line after closing bracket, it's a bit cramped.

**Review by rlubos — COMMENTED** [review]
Just a few nits, but looks good overall.

**Review by sigvartmh — APPROVED** [review]
Fix the nits and it looks good to me too.

File: include/net/fota_download.h (hunk 1 above)
Comment (junqingzou) [reply include/net/fota_download.h]: Added

File: subsys/net/lib/fota_download/src/fota_download.c (hunk 3 above)
Comment (junqingzou) [reply subsys/net/lib/fota_download/src/fota_download.c]: Added

File: subsys/net/lib/fota_download/src/fota_download.c (hunk 2 above)
Comment (junqingzou) [reply subsys/net/lib/fota_download/src/fota_download.c]: Early return added

**Review by rlubos — APPROVED** [review]

**Comment by rlubos** [comment]
@junqingzou There are valid CI failures, please have a look:"

> /home/work/nrf/subsys/net/lib/aws_fota/src/aws_fota.c:550:2: error: enumeration value 'FOTA_DOWNLOAD_EVT_CANCELLED' not handled in switch [-Werror=switch]
  550 |  switch (evt->id) {
      |  ^~~~~~

**Comment by junqingzou** [comment]
> @junqingzou There are valid CI failures, please have a look:
> 
> > /home/work/nrf/subsys/net/lib/aws_fota/src/aws_fota.c:550:2: error: enumeration value 'FOTA_DOWNLOAD_EVT_CANCELLED' not handled in switch [-Werror=switch]
> > 550 |  switch (evt->id) {
> > |  ^~~~~~

Thanks @rlubos , there is no "default" in "switch" there and I just added one.

**Pull request merged**